timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/BatchingPaxosLatestSequenceCache.java

```diff
@@ -16,92 +16,159 @@
+            throws InvalidAcceptorCacheKeyException {
+        AcceptorCacheDigest digest = delegate.latestSequencesPreparedOrAccepted(Optional.empty(), clientsSeenSoFar);
+        ConcurrentMap<Client, PaxosLong> newEntriesToCache =
+                cacheKeysToCaches.get(TimestampedAcceptorCacheKey.of(digest));
+        processDigest(newEntriesToCache, digest);
+        return getResponseMap(newEntriesToCache, requestedClients);
+    }
 
-        Set<Client> newClients = Sets.difference(clients, cachedEntries.keySet());
+    private Map<Client, PaxosLong> populateExistingCache(
+            TimestampedAcceptorCacheKey timestampedCacheKey,
+            ConcurrentMap<Client, PaxosLong> currentCachedEntries,
+            Set<Client> requestedClients)
+            throws InvalidAcceptorCacheKeyException {
+        Set<Client> newClients = ImmutableSet.copyOf(Sets.difference(requestedClients, currentCachedEntries.keySet()));
         if (newClients.isEmpty()) {
-            delegate.latestSequencesPreparedOrAcceptedCached(cacheKey).ifPresent(this::processDigest);
-            return getResponseMap(clients);
+            delegate.latestSequencesPreparedOrAcceptedCached(timestampedCacheKey.cacheKey())
+                    .ifPresent(digest -> processDigest(currentCachedEntries, digest));
+            return getResponseMap(currentCachedEntries, requestedClients);
         } else {
-            processDigest(delegate.latestSequencesPreparedOrAccepted(Optional.of(cacheKey), newClients));
-            return getResponseMap(clients);
+            processDigest(currentCachedEntries, delegate.latestSequencesPreparedOrAccepted(
+                    Optional.of(timestampedCacheKey.cacheKey()),
+                    newClients));
+            return getResponseMap(currentCachedEntries, requestedClients);
+        }
+    }
+
+    private void processDigest(ConcurrentMap<Client, PaxosLong> currentCachedEntries, AcceptorCacheDigest digest) {
+        TimestampedAcceptorCacheKey newCacheKey = TimestampedAcceptorCacheKey.of(digest);
+        // this shares the same map with "previous" cache keys, if it's too confusing we can always copy it potentially
+        ConcurrentMap<Client, PaxosLong> newCachedEntries =
+                cacheKeysToCaches.get(newCacheKey, $ -> currentCachedEntries);
+        KeyedStream.stream(digest.updates())
+                .map(PaxosLong::of)
+                .forEach((client, paxosLong) ->
+                        newCachedEntries.merge(client, paxosLong, BatchingPaxosLatestSequenceCache::max));
+
+        // for a *new* mapping, setting the cache key must happen *after* we've setup the mapping, so that concurrent
+        // clients will not reference an in-progress populating map which can be empty.
+        maybeSetNewCacheKey(newCacheKey);
+    }
+
+    private void maybeSetNewCacheKey(TimestampedAcceptorCacheKey newCacheKey) {
+        while (true) {
+            TimestampedAcceptorCacheKey current = cacheKey.get();
```

Comment:
should I have passed in the cache key that generated this request? I think it's probably fine, but might make sense